### PR TITLE
FFN layer registry

### PR DESCRIPTION
**Pull request by dakinggg**
Before and after
Screenshot 2024-04-04 at 7 14 24 PM
Screenshot 2024-04-04 at 7 15 23 PM

### CODE REVIEW

File: llmfoundry/registry.py [hunk 1]
```diff
@@ -119,4 +120,5 @@
     'models',
     'metrics',
     'dataloaders',
+    'norms',
```
Comment (dakinggg) [inline llmfoundry/registry.py:123]: add

File: tests/test_registry.py [hunk 2]
```diff
@@ -30,6 +30,7 @@ def test_expected_registries_exist():
         'dataloaders',
         'metrics',
         'models',
+        'norms',
```
Comment (dakinggg) [inline tests/test_registry.py:33]: add

**Review by mvpatel2000 — COMMENTED** [review]
will review after tests pass

**Comment by dakinggg** [comment]
And with the MB merge:

Screenshot 2024-04-11 at 8 24 30 PM
Screenshot 2024-04-11 at 8 24 40 PM

@irenedea @mvpatel2000 this one is ready

**Comment by dakinggg** [comment]
jk, also failed a test

**Comment by dakinggg** [comment]
ok @mvpatel2000 tests passing now

**Review by mvpatel2000 — COMMENTED** [review]
why do the megablocks ones have to be separate?

@vchiley to review the mb build stuff

**Comment by dakinggg** [comment]
@mvpatel2000 because there are a bunch of places in the code that we do something only for the mb layers (e.g. the mb args as an attr on the main class.

File: llmfoundry/models/layers/ffn.py [hunk 3]
```diff
@@ -200,131 +225,225 @@ def build_ffn(
...
+                                                ffn_hidden_size),
+        moe_num_experts=moe_num_experts,
+        moe_top_k=moe_top_k,
+        mlp_type=mlp_type,
+        bias=bias,
+        moe_jitter_eps=moe_jitter_eps,
+        activation_fn=resolve_ffn_act_fn(ffn_act_fn),
+        moe_normalize_expert_weights=moe_normalize_expert_weights,
+        uniform_expert_assignment=uniform_expert_assignment,
+        device=device,  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (mvpatel2000) [inline llmfoundry/models/layers/ffn.py:277]: Why ignore? I think we can remove this 🤔

Comment (mvpatel2000) [reply llmfoundry/models/layers/ffn.py:277]: hm nvm I think @bigning has a PR for this

**Review by mvpatel2000 — APPROVED** [review]

File: llmfoundry/models/layers/ffn.py (hunk 3 above)
Comment (dakinggg) [reply llmfoundry/models/layers/ffn.py:277]: Ah ok, I'll try. Was just a copy paste from existing code.

Comment (dakinggg) [reply llmfoundry/models/layers/ffn.py:277]: ok fixed

**Pull request merged**